ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,312 @@
+  private final int initialDelay;
+  private final int intervalPeriod;
+
+
+  public TransientQueryCleanupService(final ServiceContext serviceContext,
+                                      final KsqlConfig ksqlConfig) {
+    this.transientPattern = Pattern.compile(
+            ksqlConfig.getString(
+                    KsqlConfig.KSQL_TRANSIENT_QUERY_NAME_PREFIX_CONFIG));
+
+    this.initialDelay = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_INITIAL_DELAY_SECONDS);
+
+    this.intervalPeriod = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_PERIOD_SECONDS);
+
+    topicsCleanupTasks = new LinkedBlockingDeque<>();
+    stateDirsCleanupTasks = new LinkedBlockingDeque<>();
+
+    this.stateDir = ksqlConfig.getKsqlStreamConfigProps()
+            .getOrDefault(
+                    StreamsConfig.STATE_DIR_CONFIG,
+                    StreamsConfig.configDef()
+                            .defaultValues()
+                            .get(StreamsConfig.STATE_DIR_CONFIG))
+            .toString();
+
+    this.serviceContext = serviceContext;
+    this.isLocalCommandsInitialized = false;
+    this.isLocalCommandsProcessed = false;
+  }
+
+  @Override
+  protected void runOneIteration() {
+    if (isLocalCommandsInitialized && !isLocalCommandsProcessed) {
+      LOG.info("Adding LocalCommands to TransientQueryCleanupService.");
+      localCommandsQueryAppIds.forEach(id -> {
+        addTopicCleanupTask(new TransientQueryTopicCleanupTask(serviceContext, id));
+        addStateCleanupTask(new TransientQueryStateCleanupTask(id, stateDir));
+      });
+      isLocalCommandsProcessed = true;
+    }
+
+    LOG.info("Starting cleanup for leaked resources.");
+
+    topicsCleanupTasks.addAll(findPossiblyLeakedTransientTopics());
+    stateDirsCleanupTasks.addAll(findPossiblyLeakedStateDirs());
+
+    final int numLeakedTopics = topicsCleanupTasks.size();
+    final int numLeakedStateDirs = stateDirsCleanupTasks.size();
+
+    LOG.info("Cleaning up {} leaked topics.", numLeakedTopics);
+    cleanupLeakedTopics();
+
+    LOG.info("Cleaning up {} leaked state directories.", numLeakedStateDirs);
+    cleanupLeakedStateDirs();
+  }
+
+  @Override
+  public Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(initialDelay, intervalPeriod, TimeUnit.SECONDS);
+  }
+
+  public void cleanupLeakedTopics() {
+    try {
+      while (!topicsCleanupTasks.isEmpty()) {
+        final Callable<Boolean> task = topicsCleanupTasks.take();
+
+        try {
+          final boolean cleanedUp = task.call();
+          if (!cleanedUp) {
+            topicsCleanupTasks.add(task);
+          }
+        } catch (Exception e) {
+          topicsCleanupTasks.add(task);
+        }
+      }
+    } catch (final InterruptedException e) {
+      // gracefully exit if this method was interrupted and reset
+      // the interrupt flag
+      Thread.currentThread().interrupt();
+    }
```

Comment:
It does seem like a bummer that we don't just collect all the topic names and send a single delete-topics request (Alan also pointed this out below).

I'm fine if you don't want to refactor it; just calling this out.

Reply:
Also, even though the internal KSQL API might be synchonous, the actual Admin client call is async. So you could just fire off the async request and immediately move on to cleaning up the state directories. Then, after the state directory deletion is done, you can wait on the Future from the delete-topics so that you can actually log whether it succeeded or failed, and also to prevent stacking up async calls on the brokers.

Reply:
i agree. Refactored and changed send a single delete-topics batch request.